src/render/md.jl

```diff
@@ -43,50 +56,84 @@ function writemd(io::IO, doc::Metadata, config::Config)
+                                   push!(API_INDEX.modulenames, string(modulename(doc)));
+                                   push!(API_INDEX.entriesmd, ents))
+            println(io)
+            process_entriesmd(io, "Exported", ents.exported, config)
+            process_entriesmd(io, "Internal", ents.internal, config)
+        end
+    end
+end
+
+function prepare_entriesmd(index::Dict{Symbol, Any}, doc::Metadata, config::Config)
+    ents = EntriesMD(config)
+    if (config.md_genindex || config.md_permalink)
+        pageanchors = Dict{Symbol, Dict{String, Int}}([])
+        for k in config.category_order
             haskey(index, k) || continue
+            k in pageanchors || (pageanchors[k] = Dict([]))
+            basenames = pageanchors[k]
             for (s, obj) in index[k]
-                push!(ents, modulename(doc), obj, entries(doc)[obj])
+                ent = entries(doc)[obj]
+                if k == :comment
+                    basename = "comment"
+                else
+                    basename = k == :macro                                          ?
+                            Base.lowercase(string(Docile.Interface.macroname(ent))) :
+                            Base.lowercase(string(Docile.Interface.name(obj)))
+                end
+                basename in keys(basenames)                  ?
+                        anchornum = basenames[basename] += 1 :
+                        anchornum = basenames[basename] = 1
+                push!(ents, modulename(doc), obj, entries(doc)[obj],
+                        "$(string(k))__$(basename).$(anchornum)", k)
+            end
+        end
+    else
+        for k in config.category_order
+            haskey(index, k) || continue
+            for (s, obj) in index[k]
+                push!(ents, modulename(doc), obj, entries(doc)[obj], "", k)
             end
         end
-        println(io)
-        writemd(io, ents, config)
     end
+    return ents
 end
 
-function writemd(io::IO, ents::Entries, config::Config)
-    exported = Entries()
-    internal = Entries()
-
-    for (modname, obj, ent) in ents.entries
-        isexported(modname, obj) ?
-            push!(exported, modname, obj, ent) :
-            config.include_internal && push!(internal, modname, obj, ent)
-    end
-
-    if !isempty(exported.entries)
-        print_help(io, config.mdstyle_exported, "Exported")
-        for (modname, obj, ent) in exported.entries
-            writemd(io, modname, obj, ent, config)
-        end
-    end
-    if !isempty(internal.entries)
-        print_help(io, config.mdstyle_internal, "Internal")
-        for (modname, obj, ent) in internal.entries
-            writemd(io, modname, obj, ent, config)
+function process_entriesmd(io::IO, grpname::ByteString,
+                           entries::Dict{Symbol, Vector{(Module, Any, Entry, ByteString)}},
+                           config::Config)
```

Comment:
`ByteString` -> `AbstractString`.

Just `entries::Dict` rather that longer type, shouldn't be any method ambiguities by doing that.